foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/descriptors/ClassDescriptor.java

```diff
@@ -57,6 +57,7 @@
 import java.util.Map;
 import java.util.Set;
 import java.util.Vector;
+import java.util.concurrent.ConcurrentHashMap;
```

Comment:
in most cases, usage of concurrent collections is not needed, usually the fix for concurrentmodificationexception is to fix the loop, which modifies the content of the collection being iterated instead. Do we have a stacktrace which may help identifying that place/loop?

Reply:
Stacktrace :
`java.lang.AssertionError: Exception was caught:
java.util.concurrent.ExecutionException:
java.util.ConcurrentModificationException at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
 at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
  at org.eclipse.persistence.jpa.test.mapping.TestMultitenantOneToMany.testMultitenancySchemaDescriminatorWithOneToMany(TestMultitenantOneToMany.java:105)
   at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
     at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
      at java.base/java.lang.reflect.Method.invoke(Method.java:566)
       at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
         at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
          at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
           at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
            at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
             at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
              at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
               at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
                at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
                 at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
                  at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
                   at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
                    at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
                     at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
                      at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
                       at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
                        at org.eclipse.persistence.jpa.test.framework.EmfRunner.run(EmfRunner.java:43)
                         at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
                          at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
                           at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
                            at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
                             at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:377)
                              at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:138)
                               at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:465)
                                at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:451)
                                 Caused by: java.util.ConcurrentModificationException
                                  at java.base/java.util.HashMap$HashIterator.nextNode(HashMap.java:1493)
                                   at java.base/java.util.HashMap$KeyIterator.next(HashMap.java:1516)
                                    at org.eclipse.persistence.descriptors.ClassDescriptor.notifyReferencingDescriptorsOfIsolation(ClassDescriptor.java:4123)
                                     at org.eclipse.persistence.descriptors.CachePolicy.postInitialize(CachePolicy.java:193)
                                      at org.eclipse.persistence.descriptors.ClassDescriptor.postInitialize(ClassDescriptor.java:4052)
                                       at org.eclipse.persistence.internal.sessions.AbstractSession.updateTablePerTenantDescriptors(AbstractSession.java:1427)
                                        at org.eclipse.persistence.sessions.server.ClientSession.<init>(ClientSession.java:150)
                                         at org.eclipse.persistence.sessions.server.ServerSession.acquireClientSession(ServerSession.java:432)
                                          at org.eclipse.persistence.internal.jpa.EntityManagerImpl.getActivePersistenceContext(EntityManagerImpl.java:2087)
                                           at org.eclipse.persistence.internal.jpa.EntityManagerImpl.getActiveSession(EntityManagerImpl.java:1358)
                                            at org.eclipse.persistence.internal.jpa.EntityManagerImpl.getQueryHints(EntityManagerImpl.java:2750
                                             at org.eclipse.persistence.internal.jpa.EntityManagerImpl.find(EntityManagerImpl.java:696)
                                              at org.eclipse.persistence.jpa.test.mapping.TestMultitenantOneToMany.load(TestMultitenantOneToMany.java:143)
                                               at org.eclipse.persistence.jpa.test.mapping.TestMultitenantOneToMany.execute(TestMultitenantOneToMany.java:129)
                                                at org.eclipse.persistence.jpa.test.mapping.TestMultitenantOneToMany.lambda$testMultitenancySchemaDescriminatorWithOneToMany$1(TestMultitenantOneToMany.java:95)
                                                 at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
                                                  at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
                                                   at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
                                                    at java.base/java.lang.Thread.run(Thread.java:834)
                                                     at org.eclipse.persistence.jpa.test.mapping.TestMultitenantOneToMany.testMultitenancySchemaDescriminatorWithOneToMany(TestMultitenantOneToMany.java:114)`

Reply:
Great! I got deeper and this `ConcurrentModificationException` is caused by a bug in  [ClassDescriptor.clone()](https://github.com/eclipse-ee4j/eclipselink/blob/3.0.1/foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/descriptors/ClassDescriptor.java#L1371), which resets `referencingClasses` field on the current instance, it should copy current content to cloned instance instead. There is also [older bug](https://bugs.eclipse.org/bugs/show_bug.cgi?id=463737), which has a patch (I'd only avoid adding new private method as is in the suggested patch)